### PR TITLE
tests: generalise ui testing and allow for per file configuration

### LINKED ISSUE #460: tests: generalise testing infrastructure to run tests for any stage

### PR DESCRIPTION
**Pull request by feds01**
This patch makes the testing infrastructure more generalised and available to any of the stages within the compiler. This change shifts tests from using a configuration style that's based on the name of the directory, to what the file specifies in it's
own configuration. 

Test files can now specify their configuration parameters within the file itself on the first line of the file. This configuration is a simple key-pair set that produces a `TestMetadata` that is generated at build time and processed within `tests/testing-internal/src/metadata.rs`. An example of this configuration would be:

```rs
// run=fail, stage=semantic

main := () => {
    [...x, t, ...] := a;    

    match t {
        [...a, [...l, t, ..., c], t] => {};
        _ => {};
    };
};
```
This case specifies that the test should run up to `semantics` and that it should `fail`. 

With this kind of configuration, it is easily possible to expand it and make it far more sophisticated with all kinds of possible configuration options. In light of all of this, it is now also possible to write tests for the semantics analysis, typechecking, exhaustiveness... and any other desired stage.

closes #460

### CODE REVIEW

**Comment by kontheocharis** [comment]
Very nice sir

File: tests/testing-internal/src/metadata.rs [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+}
+
+impl TestMetadataBuilder {
+    /// Create a new [FileMetadataBuilder]
+    pub fn new() -> Self {
+        Self { stage: None, completion: None }
+    }
+
+    /// Add a stage value to the test.
+    pub fn with_stage(&mut self, stage: CompilerMode) -> &mut Self {
```
[the remark this answers is not in the document]
Comment (feds01) [reply tests/testing-internal/src/metadata.rs:94]: This isn't what `ReportBuilder` pattern is using

File: tests/runner.rs [hunk 2]
```diff
@@ -0,0 +1,154 @@
+//! Hash Compiler UI Test runner, this file will read the `cases` directory for
+//! `.hash` files in the specific `should_pass` and `should_fail` directories.
```
[the remark this answers is not in the document]
Comment (feds01) [reply tests/runner.rs:2]: Yes, but they now become irrelevant, in a subsequent commit, I updated this description

File: tests/testing-internal/src/metadata.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (feds01) [reply tests/testing-internal/src/metadata.rs:94]: Also, this isn't the case for the rust example: https://doc.rust-lang.org/1.0.0/style/ownership/builders.html

[the remark this answers is not in the document]
Comment (kontheocharis) [reply tests/testing-internal/src/metadata.rs:87]: I've been bamboozled

**Review by kontheocharis — APPROVED** [review]

**Pull request merged**